File: Java/CodesInterviews/src/main/java/org/example/arrays/ProductOfArray.java
package org.example.arrays;

import java.util.Arrays;

public class ProductOfArray {

    public int[] productExceptSelf(int[] nums) {
        int[] prefixProduct = new int[nums.length];
        prefixProduct[0] = 1;

        for (int i = 1; i < nums.length; i++) {
            prefixProduct[i] = prefixProduct[i-1] * nums[i-1];
        }

        int suffixProduct = 1;
        for (int i = nums.length-1; i >= 0; i--) {
            prefixProduct[i] = prefixProduct[i] * suffixProduct;
            suffixProduct = suffixProduct * nums[i];
        }

        return prefixProduct;
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 4};
        ProductOfArray product = new ProductOfArray();
        System.out.println(Arrays.toString(product.productExceptSelf(nums)));
    }

}

File: Kotlin/CodesInterviews/src/main/kotlin/arrays/ProductOfArray.kt
package arrays

class ProductOfArray {

    fun productExceptSelf(nums: IntArray): IntArray {
        val n = nums.size
        val prefixProduct = IntArray(n)
        prefixProduct[0] = 1

        for (i in 1..<n) {
            prefixProduct[i] = prefixProduct[i - 1] * nums[i - 1]
        }

        var suffixProduct = 1
        for (i in nums.indices.reversed()) {
            prefixProduct[i] =  prefixProduct[i] * suffixProduct
            suffixProduct *= nums[i]
        }

        return prefixProduct
    }
}

fun main() {
    val product = ProductOfArray()
    val nums = intArrayOf(1, 2, 3, 4)
    val nums2 = intArrayOf(11, 13, 12, 34)
    print(product.productExceptSelf(nums).contentToString())
    print(product.productExceptSelf(nums2).contentToString())
}